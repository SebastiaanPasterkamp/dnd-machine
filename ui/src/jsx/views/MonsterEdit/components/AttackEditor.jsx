import React from 'react';
import PropTypes from 'prop-types';
import {
    get,
} from 'lodash/fp';

import { memoize } from '../../../utils';

import ListDataWrapper from '../../../hocs/ListDataWrapper';

import DamageEdit from '../../../components/DamageEdit';
import ControlGroup from '../../../components/ControlGroup';
import InputField from '../../../components/InputField';
import ReachEdit from '../../../components/ReachEdit';
import SingleSelect from '../../../components/SingleSelect';
import { ListComponent } from '../../../components/ListComponent';
import MarkdownTextField from '../../../components/MarkdownTextField';

export class AttackEditor extends React.Component
{
    constructor(props) {
        super(props);
        this.memoize = memoize.bind(this);
    }

    onFieldChange(field) {
        const { setState } = this.props;
        return this.memoize(
            field,
            (value, callback=null) => setState(
                {[field]: value },
                callback
            )
        );
    }

    render() {
        const {
            name, description, damage, target, target_methods,
            mode, attack_modes, reach, on_hit, on_mis,
        } = this.props;
        const damageTypes = get('type', damage);

        return (
            <div className="edit-attack">
                <ControlGroup label="Name">
                    <InputField
                        placeholder="Name..."
                        data-field="name"
                        value={name}
                        setState={this.onFieldChange('name')}
                    />
                </ControlGroup>
                <ControlGroup label="Description">
                    <MarkdownTextField
                        placeholder="Description..."
                        value={description}
                        rows={5}
                        setState={this.onFieldChange('description')}
                    />
                </ControlGroup>
                <ListComponent
                    list={damage}
                    component={DamageEdit}
                    newItem="initial"
                    keyProp="type"
                    setState={this.onFieldChange('damage')}
                    componentProps={{
                        disabledTypes: damageTypes
                    }}
                />
                <ControlGroup label="Target">
                    <SingleSelect
                        emptyLabel="Target..."
                        selected={target}
                        items={target_methods}
                        setState={this.onFieldChange('target')}
                    />
                </ControlGroup>
                <ControlGroup label="Mode">
                    <SingleSelect
                        emptyLabel="Mode..."
                        selected={mode}
                        items={attack_modes}
                        setState={this.onFieldChange('mode')}
                    />
                </ControlGroup>
                <ReachEdit
                    {...reach}
                    setState={this.onFieldChange('reach')}
                />
                <ControlGroup label="On Hit">
                    <MarkdownTextField
                        className="small"
                        placeholder="Bad stuff..."
                        value={on_hit}
                        rows={5}
                        setState={this.onFieldChange('on_hit')}
                    />
                </ControlGroup>
                <ControlGroup label="On Mis">
                    <MarkdownTextField
                        className="small"
                        placeholder="Still bad stuff..."
                        value={on_mis}
                        rows={5}
                        setState={this.onFieldChange('on_mis')}
                    />
                </ControlGroup>
            </div>
        );
    }
}

AttackEditor.propTypes = {
    name: PropTypes.string,
    description: PropTypes.string,
    damage: PropTypes.array,
    target: PropTypes.string,
    target_methods: PropTypes.array,
    mode: PropTypes.string,
    attack_modes: PropTypes.array,
    reach: PropTypes.object,
    on_hit: PropTypes.string,
    on_mis: PropTypes.string,
    setState: PropTypes.func.isRequired,
};

AttackEditor.defaultProps = {
    name: '',
    description: '',
    damage: [],
    target: '',
    target_methods: [],
    mode: '',
    attack_modes: [],
    reach: {},
    on_hit: '',
    on_mis: '',
};

export default ListDataWrapper(
    AttackEditor,
    [
        'target_methods',
        'attack_modes',
    ],
    'items'
);
