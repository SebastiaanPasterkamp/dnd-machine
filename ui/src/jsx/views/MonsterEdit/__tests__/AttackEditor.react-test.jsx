import React from 'react';
import {mount} from 'enzyme';
import renderer from 'react-test-renderer';

import {mockedApi, attack_modes, damage_types, target_methods} from '../../../../../tests/__mocks__';

import AttackEditor from '../components/AttackEditor';

describe('AttackEditor', () => {
    const fullProps = {
        name: 'Punch',
        description: 'Fist to Face',
        damage: [
            {
                dice_count: 1,
                dice_size: 4,
                bonus: 1,
                type: 'bludgeoning'
            }
        ],
        target: 'single',
        mode: 'melee',
        reach: {
            min: 5,
            max: 5
        },
        on_hit: "It hurts",
        on_mis: "It doesn't hurt"
    };

    beforeAll(() => {
        fetch.mockImplementation(mockedApi({
            attack_modes,
            damage_types,
            target_methods,
        }));
    });

    afterAll(() => {
        fetch.resetMocks();
    });

    it('should render without props', () => {
        const setState = jest.fn();
        const tree = renderer.create(<AttackEditor setState={setState}/>);

        expect(tree).toMatchSnapshot();
    });

    it('should render with full props', () => {
        const setState = jest.fn();

        const tree = renderer.create(<AttackEditor {...fullProps} setState={setState}/>);

        expect(tree).toMatchSnapshot();
    });

    describe('while editing', () => {
        const setState = jest.fn();
        let wrapper;

        beforeAll(() => {
            wrapper = mount(<AttackEditor setState={setState}/>);
        });

        it('should accept a name change', () => {
            wrapper.find('input[data-field="name"]').simulate('change', {
                target: {
                    value: fullProps.name
                }
            });

            expect(setState).toBeCalledWith({
                name: fullProps.name
            }, null)
        });

        it('should accept a target change', () => {
            wrapper.find('button').at(0).simulate('click');
            wrapper.find('li[data-value="area"]').simulate('click');

            expect(setState).toBeCalledWith({
                target: 'area'
            }, null)
        });
    });
});
