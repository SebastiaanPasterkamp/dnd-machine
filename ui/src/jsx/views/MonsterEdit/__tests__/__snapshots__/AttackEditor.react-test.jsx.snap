// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttackEditor should render with full props 1`] = `
<div
  className="edit-attack"
>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Name
    </span>
    <input
      className="nice-form-control"
      data-field="name"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Name..."
      type={undefined}
      value="Punch"
    />
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Description
    </span>
    <div
      className="markdown-textedit preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Description..."
        rows={5}
        value="Fist to Face"
      />
      <span
        className="markdown-textedit__preview"
      >
        <p>
          Fist to Face
        </p>
      </span>
    </div>
  </div>
  <ul
    className="list-component"
  >
    <li>
      <div
        className="nice-control-group"
      >
        <span
          className="nice-input-addon"
        >
          Damage
        </span>
        <div
          className="nice-dropdown nice-form-control"
        >
          <button
            className="nice-btn"
            onClick={[Function]}
          >
            1
            <i
              className="icon fa-angle-down"
            />
          </button>
          <ul
            className="dropdown-menu"
            onClick={[Function]}
          >
            <li
              className={null}
              data-value={0}
              onClick={[Function]}
            >
              <a>
                0
              </a>
            </li>
            <li
              className="info"
              data-value={1}
              onClick={[Function]}
            >
              <a>
                1
              </a>
            </li>
            <li
              className={null}
              data-value={2}
              onClick={[Function]}
            >
              <a>
                2
              </a>
            </li>
            <li
              className={null}
              data-value={3}
              onClick={[Function]}
            >
              <a>
                3
              </a>
            </li>
            <li
              className={null}
              data-value={4}
              onClick={[Function]}
            >
              <a>
                4
              </a>
            </li>
            <li
              className={null}
              data-value={5}
              onClick={[Function]}
            >
              <a>
                5
              </a>
            </li>
            <li
              className={null}
              data-value={6}
              onClick={[Function]}
            >
              <a>
                6
              </a>
            </li>
            <li
              className={null}
              data-value={7}
              onClick={[Function]}
            >
              <a>
                7
              </a>
            </li>
            <li
              className={null}
              data-value={8}
              onClick={[Function]}
            >
              <a>
                8
              </a>
            </li>
            <li
              className={null}
              data-value={9}
              onClick={[Function]}
            >
              <a>
                9
              </a>
            </li>
            <li
              className={null}
              data-value={10}
              onClick={[Function]}
            >
              <a>
                10
              </a>
            </li>
            <li
              className={null}
              data-value={11}
              onClick={[Function]}
            >
              <a>
                11
              </a>
            </li>
            <li
              className={null}
              data-value={12}
              onClick={[Function]}
            >
              <a>
                12
              </a>
            </li>
            <li
              className={null}
              data-value={13}
              onClick={[Function]}
            >
              <a>
                13
              </a>
            </li>
            <li
              className={null}
              data-value={14}
              onClick={[Function]}
            >
              <a>
                14
              </a>
            </li>
            <li
              className={null}
              data-value={15}
              onClick={[Function]}
            >
              <a>
                15
              </a>
            </li>
            <li
              className={null}
              data-value={16}
              onClick={[Function]}
            >
              <a>
                16
              </a>
            </li>
            <li
              className={null}
              data-value={17}
              onClick={[Function]}
            >
              <a>
                17
              </a>
            </li>
            <li
              className={null}
              data-value={18}
              onClick={[Function]}
            >
              <a>
                18
              </a>
            </li>
            <li
              className={null}
              data-value={19}
              onClick={[Function]}
            >
              <a>
                19
              </a>
            </li>
          </ul>
        </div>
        <span
          className="nice-input-addon"
        >
          d
        </span>
        <div
          className="nice-dropdown nice-form-control"
        >
          <button
            className="nice-btn"
            onClick={[Function]}
          >
            4
            <i
              className="icon fa-angle-down"
            />
          </button>
          <ul
            className="dropdown-menu"
            onClick={[Function]}
          >
            <li
              className="info"
              data-value={4}
              onClick={[Function]}
            >
              <a>
                4
              </a>
            </li>
            <li
              className={null}
              data-value={6}
              onClick={[Function]}
            >
              <a>
                6
              </a>
            </li>
            <li
              className={null}
              data-value={8}
              onClick={[Function]}
            >
              <a>
                8
              </a>
            </li>
            <li
              className={null}
              data-value={10}
              onClick={[Function]}
            >
              <a>
                10
              </a>
            </li>
            <li
              className={null}
              data-value={12}
              onClick={[Function]}
            >
              <a>
                12
              </a>
            </li>
          </ul>
        </div>
        <span
          className="nice-input-addon"
        >
          +
        </span>
        <input
          className="nice-form-control"
          disabled={undefined}
          onChange={[Function]}
          onKeyPress={null}
          placeholder="Bonus..."
          type={undefined}
          value={1}
        />
        <span
          className="nice-input-addon"
        >
          Type
        </span>
        <div
          className="nice-dropdown nice-form-control"
        >
          <button
            className="nice-btn"
            onClick={[Function]}
          >
            Bludgeoning
            <i
              className="icon fa-angle-down"
            />
          </button>
          <ul
            className="dropdown-menu"
            onClick={[Function]}
          >
            <li
              className={null}
              data-value="acid"
              onClick={[Function]}
            >
              <a>
                Acid
              </a>
            </li>
            <li
              className="info"
              data-value="bludgeoning"
              onClick={[Function]}
            >
              <a>
                Bludgeoning
              </a>
            </li>
            <li
              className={null}
              data-value="cold"
              onClick={[Function]}
            >
              <a>
                Cold
              </a>
            </li>
            <li
              className={null}
              data-value="fire"
              onClick={[Function]}
            >
              <a>
                Fire
              </a>
            </li>
            <li
              className={null}
              data-value="force"
              onClick={[Function]}
            >
              <a>
                Force
              </a>
            </li>
            <li
              className={null}
              data-value="lightning"
              onClick={[Function]}
            >
              <a>
                Lightning
              </a>
            </li>
            <li
              className={null}
              data-value="necrotic"
              onClick={[Function]}
            >
              <a>
                Necrotic
              </a>
            </li>
            <li
              className={null}
              data-value="piercing"
              onClick={[Function]}
            >
              <a>
                Piercing
              </a>
            </li>
            <li
              className={null}
              data-value="poison"
              onClick={[Function]}
            >
              <a>
                Poison
              </a>
            </li>
            <li
              className={null}
              data-value="psychic"
              onClick={[Function]}
            >
              <a>
                Psychic
              </a>
            </li>
            <li
              className={null}
              data-value="radiant"
              onClick={[Function]}
            >
              <a>
                Radiant
              </a>
            </li>
            <li
              className={null}
              data-value="slashing"
              onClick={[Function]}
            >
              <a>
                Slashing
              </a>
            </li>
            <li
              className={null}
              data-value="thunder"
              onClick={[Function]}
            >
              <a>
                Thunder
              </a>
            </li>
          </ul>
        </div>
      </div>
      <button
        className="nice-btn list-component__delete warning icon fa-minus"
        name="del"
        onClick={[Function]}
      >
        ​
      </button>
    </li>
    <li>
      <button
        className="nice-btn list-component__add primary icon fa-plus"
        name="add"
        onClick={[Function]}
      >
        ​
      </button>
    </li>
  </ul>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Target
    </span>
    <div
      className="nice-dropdown nice-form-control"
    >
      <button
        className="nice-btn"
        onClick={[Function]}
      >
        Single target
        <i
          className="icon fa-angle-down"
        />
      </button>
      <ul
        className="dropdown-menu"
        onClick={[Function]}
      >
        <li
          className="info"
          data-value="single"
          onClick={[Function]}
        >
          <a>
            Single target
          </a>
        </li>
        <li
          className={null}
          data-value="multiple"
          onClick={[Function]}
        >
          <a>
            Multi-target
          </a>
        </li>
        <li
          className={null}
          data-value="area"
          onClick={[Function]}
        >
          <a>
            Area of Effect
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Mode
    </span>
    <div
      className="nice-dropdown nice-form-control"
    >
      <button
        className="nice-btn"
        onClick={[Function]}
      >
        Melee Attack
        <i
          className="icon fa-angle-down"
        />
      </button>
      <ul
        className="dropdown-menu"
        onClick={[Function]}
      >
        <li
          className="info"
          data-value="melee"
          onClick={[Function]}
        >
          <a>
            Melee Attack
          </a>
        </li>
        <li
          className={null}
          data-value="ranged"
          onClick={[Function]}
        >
          <a>
            Ranged Attack
          </a>
        </li>
        <li
          className={null}
          data-value="spell"
          onClick={[Function]}
        >
          <a>
            Spell Attack
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Reach
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Normal..."
      type="number"
      value={5}
    />
    <span
      className="nice-input-addon"
    >
      ft.
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Disadvantage..."
      type="number"
      value={5}
    />
    <span
      className="nice-input-addon"
    >
      ft.
    </span>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      On Hit
    </span>
    <div
      className="markdown-textedit small preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Bad stuff..."
        rows={5}
        value="It hurts"
      />
      <span
        className="markdown-textedit__preview"
      >
        <p>
          It hurts
        </p>
      </span>
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      On Mis
    </span>
    <div
      className="markdown-textedit small preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Still bad stuff..."
        rows={5}
        value="It doesn't hurt"
      />
      <span
        className="markdown-textedit__preview"
      >
        <p>
          It doesn't hurt
        </p>
      </span>
    </div>
  </div>
</div>
`;

exports[`AttackEditor should render without props 1`] = `
<div
  className="edit-attack"
>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Name
    </span>
    <input
      className="nice-form-control"
      data-field="name"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Name..."
      type={undefined}
      value=""
    />
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Description
    </span>
    <div
      className="markdown-textedit preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="markdown-textedit__placeholder"
      >
        Description...
      </span>
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Description..."
        rows={5}
        value=""
      />
      <span
        className="markdown-textedit__preview"
      />
    </div>
  </div>
  <ul
    className="list-component"
  >
    <li>
      <div
        className="nice-control-group"
      >
        <span
          className="nice-input-addon"
        >
          Damage
        </span>
        <div
          className="nice-dropdown nice-form-control"
        >
          <button
            className="nice-btn"
            onClick={[Function]}
          >
            <i
              className="icon fa-angle-down"
            />
          </button>
          <ul
            className="dropdown-menu"
            onClick={[Function]}
          >
            <li
              className={null}
              data-value={0}
              onClick={[Function]}
            >
              <a>
                0
              </a>
            </li>
            <li
              className={null}
              data-value={1}
              onClick={[Function]}
            >
              <a>
                1
              </a>
            </li>
            <li
              className={null}
              data-value={2}
              onClick={[Function]}
            >
              <a>
                2
              </a>
            </li>
            <li
              className={null}
              data-value={3}
              onClick={[Function]}
            >
              <a>
                3
              </a>
            </li>
            <li
              className={null}
              data-value={4}
              onClick={[Function]}
            >
              <a>
                4
              </a>
            </li>
            <li
              className={null}
              data-value={5}
              onClick={[Function]}
            >
              <a>
                5
              </a>
            </li>
            <li
              className={null}
              data-value={6}
              onClick={[Function]}
            >
              <a>
                6
              </a>
            </li>
            <li
              className={null}
              data-value={7}
              onClick={[Function]}
            >
              <a>
                7
              </a>
            </li>
            <li
              className={null}
              data-value={8}
              onClick={[Function]}
            >
              <a>
                8
              </a>
            </li>
            <li
              className={null}
              data-value={9}
              onClick={[Function]}
            >
              <a>
                9
              </a>
            </li>
            <li
              className={null}
              data-value={10}
              onClick={[Function]}
            >
              <a>
                10
              </a>
            </li>
            <li
              className={null}
              data-value={11}
              onClick={[Function]}
            >
              <a>
                11
              </a>
            </li>
            <li
              className={null}
              data-value={12}
              onClick={[Function]}
            >
              <a>
                12
              </a>
            </li>
            <li
              className={null}
              data-value={13}
              onClick={[Function]}
            >
              <a>
                13
              </a>
            </li>
            <li
              className={null}
              data-value={14}
              onClick={[Function]}
            >
              <a>
                14
              </a>
            </li>
            <li
              className={null}
              data-value={15}
              onClick={[Function]}
            >
              <a>
                15
              </a>
            </li>
            <li
              className={null}
              data-value={16}
              onClick={[Function]}
            >
              <a>
                16
              </a>
            </li>
            <li
              className={null}
              data-value={17}
              onClick={[Function]}
            >
              <a>
                17
              </a>
            </li>
            <li
              className={null}
              data-value={18}
              onClick={[Function]}
            >
              <a>
                18
              </a>
            </li>
            <li
              className={null}
              data-value={19}
              onClick={[Function]}
            >
              <a>
                19
              </a>
            </li>
          </ul>
        </div>
        <span
          className="nice-input-addon"
        >
          d
        </span>
        <div
          className="nice-dropdown nice-form-control"
        >
          <button
            className="nice-btn"
            onClick={[Function]}
          >
            <i
              className="icon fa-angle-down"
            />
          </button>
          <ul
            className="dropdown-menu"
            onClick={[Function]}
          >
            <li
              className={null}
              data-value={4}
              onClick={[Function]}
            >
              <a>
                4
              </a>
            </li>
            <li
              className={null}
              data-value={6}
              onClick={[Function]}
            >
              <a>
                6
              </a>
            </li>
            <li
              className={null}
              data-value={8}
              onClick={[Function]}
            >
              <a>
                8
              </a>
            </li>
            <li
              className={null}
              data-value={10}
              onClick={[Function]}
            >
              <a>
                10
              </a>
            </li>
            <li
              className={null}
              data-value={12}
              onClick={[Function]}
            >
              <a>
                12
              </a>
            </li>
          </ul>
        </div>
        <span
          className="nice-input-addon"
        >
          +
        </span>
        <input
          className="nice-form-control"
          disabled={undefined}
          onChange={[Function]}
          onKeyPress={null}
          placeholder="Bonus..."
          type={undefined}
          value=""
        />
        <span
          className="nice-input-addon"
        >
          Type
        </span>
        <div
          className="nice-dropdown nice-form-control"
        >
          <button
            className="nice-btn"
            onClick={[Function]}
          >
            <i
              className="icon fa-angle-down"
            />
          </button>
          <ul
            className="dropdown-menu"
            onClick={[Function]}
          />
        </div>
      </div>
    </li>
  </ul>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Target
    </span>
    <div
      className="nice-dropdown nice-form-control"
    >
      <button
        className="nice-btn"
        onClick={[Function]}
      >
        Target...
        <i
          className="icon fa-angle-down"
        />
      </button>
      <ul
        className="dropdown-menu"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Mode
    </span>
    <div
      className="nice-dropdown nice-form-control"
    >
      <button
        className="nice-btn"
        onClick={[Function]}
      >
        Mode...
        <i
          className="icon fa-angle-down"
        />
      </button>
      <ul
        className="dropdown-menu"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      Reach
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Normal..."
      type="number"
      value=""
    />
    <span
      className="nice-input-addon"
    >
      ft.
    </span>
    <input
      className="nice-form-control"
      disabled={undefined}
      onChange={[Function]}
      onKeyPress={null}
      placeholder="Disadvantage..."
      type="number"
      value=""
    />
    <span
      className="nice-input-addon"
    >
      ft.
    </span>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      On Hit
    </span>
    <div
      className="markdown-textedit small preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="markdown-textedit__placeholder"
      >
        Bad stuff...
      </span>
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Bad stuff..."
        rows={5}
        value=""
      />
      <span
        className="markdown-textedit__preview"
      />
    </div>
  </div>
  <div
    className="nice-control-group"
  >
    <span
      className="nice-input-addon"
    >
      On Mis
    </span>
    <div
      className="markdown-textedit small preview"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <span
        className="markdown-textedit__placeholder"
      >
        Still bad stuff...
      </span>
      <textarea
        className="nice-form-control"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Still bad stuff..."
        rows={5}
        value=""
      />
      <span
        className="markdown-textedit__preview"
      />
    </div>
  </div>
</div>
`;
